File: src/main/java/com/uapp_llc/controller/TaskController.java
package com.uapp_llc.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.uapp_llc.model.Column;
import com.uapp_llc.model.Task;
import com.uapp_llc.service.ColumnService;
import com.uapp_llc.service.TaskService;

@RestController
public class TaskController {

  private final ColumnService columnService;
  private final TaskService taskService;

  @Autowired
  public TaskController(ColumnService columnService, TaskService taskService) {
    this.columnService = columnService;
    this.taskService = taskService;
  }

  @GetMapping("/projects/{projectId}/columns/{columnId}/tasks")
  public Page<Task> getAll(@PathVariable Long columnId,
                           Pageable pageable) {
    Page<Task> tasks = taskService.findAll(columnId, pageable);
    return tasks;
  }

  @PostMapping("/projects/{projectId}/columns/{columnId}/tasks")
  public Task create(@PathVariable Long projectId,
                     @PathVariable Long columnId,
                     @RequestBody Task dto) {
    Column column = columnService.find(projectId, columnId);
    Task task = taskService.create(
        column,
        dto.getName(),
        dto.getDescription()
    );
    return task;
  }

  @GetMapping("/projects/{projectId}/columns/{columnId}/tasks/{id}")
  public Task get(@PathVariable Long columnId,
                  @PathVariable Long id) {
    Task task = taskService.find(id, columnId);
    return task;
  }

  @PatchMapping("/projects/{projectId}/columns/{columnId}/tasks/{id}")
  public Task update(@PathVariable Long columnId,
                     @PathVariable Long id,
                     @RequestBody Task dto) {
    Task task = taskService.update(
        id,
        columnId,
        dto.getName(),
        dto.getDescription()
    );
    return task;
  }

  @PutMapping("/projects/{projectId}/columns/{columnId}/tasks/{id}")
  public Task attachPosition(@PathVariable Long columnId,
                             @PathVariable Long id,
                             @RequestParam Integer position) {
    Task task = taskService.changePosition(id, columnId, position);
    return task;
  }

  @PutMapping("/projects/{projectId}/columns/{columnId}/tasks/{id}")
  public Task attachColumn(@PathVariable Long projectId,
                           @PathVariable("columnId") Long actualId,
                           @PathVariable Long id,
                           @RequestParam("column") Long changeId) {
    Column change = columnService.find(changeId, projectId);
    Task task = taskService.changeColumn(id, actualId, change);
    return task;
  }

  @DeleteMapping("/projects/{projectId}/columns/{columnId}/tasks/{id}")
  public void delete(@PathVariable Long columnId,
                     @PathVariable Long id) {
    taskService.delete(id, columnId);
  }

}

File: src/main/kotlin/com/uapp_llc/controller/TaskController.kt
package com.uapp_llc.controller

import com.uapp_llc.dto.task.ContentDto
import com.uapp_llc.dto.task.MoveDto
import com.uapp_llc.dto.task.TaskDto
import com.uapp_llc.mapper.TaskMapper
import com.uapp_llc.model.Column
import com.uapp_llc.model.Task
import com.uapp_llc.service.ColumnService
import com.uapp_llc.service.TaskService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class TaskController @Autowired constructor(

    private val columnService: ColumnService,
    private val taskService: TaskService

) {

  @GetMapping("/columns/{columnId}/tasks")
  fun getAll(@PathVariable columnId: Long,
             @PageableDefault(sort = ["index"]) pageable: Pageable): Page<TaskDto> {
    val tasks: Page<Task> = taskService.findAll(columnId, pageable)
    return tasks.map(TaskMapper.INSTANCE::toDto)
  }

  @PostMapping("/columns/{columnId}/tasks")
  fun create(@PathVariable columnId: Long,
             @RequestBody dto: ContentDto): TaskDto {
    val column: Column = columnService.find(columnId)
    // Possible NPE due absent validation
    val task: Task = taskService.create(
        column,
        dto.name!!,
        dto.description!!
    )
    return TaskMapper.INSTANCE.toDto(task)
  }

  @GetMapping("/columns/{columnId}/tasks/{id}")
  fun get(@PathVariable columnId: Long,
          @PathVariable id: Long): TaskDto {
    val task: Task = taskService.find(id, columnId)
    return TaskMapper.INSTANCE.toDto(task)
  }

  @PatchMapping("/columns/{columnId}/tasks/{id}")
  fun update(@PathVariable columnId: Long,
             @PathVariable id: Long,
             @RequestBody dto: ContentDto): TaskDto {
    val task: Task = taskService.update(
        id,
        columnId,
        dto.name,
        dto.description
    )
    return TaskMapper.INSTANCE.toDto(task)
  }

  @PutMapping("/columns/{columnId}/tasks/{id}")
  fun move(@PathVariable columnId: Long,
           @PathVariable id: Long,
           @RequestBody dto: MoveDto): TaskDto {
    val newColumnId: Long? = dto.newColumnId
    val newColumn: Column? =
        if (newColumnId != null && newColumnId != columnId)
          columnService.find(newColumnId)
        else null
    val task: Task = taskService.move(
        id,
        columnId,
        newColumn,
        dto.newIndex
    )
    return TaskMapper.INSTANCE.toDto(task)
  }

  @DeleteMapping("/columns/{columnId}/tasks/{id}")
  fun delete(@PathVariable columnId: Long,
             @PathVariable id: Long) {
    taskService.delete(id, columnId)
  }

}
